### ISSUE TITLE
alphabetical order for type in trait definition also

### ISSUE THREAD

**Issue by tisonkun** [rust-lang/rustfmt#5078]
I have code snippet below:

```rust
pub trait DataValue {
    type Value;
    type Mutation;
}

impl DataValue for Int64 {
    type Mutation = Int64Mutation;
    type Value = Int64Value;
}
```

where `rustfmt` will force `type`s in the `impl` block in alphabetical order, but not for the `trait` block.

IntelliJ complains "Different impl member order from the trait" in this case.

**Comment by calebcartwright** [comment]
> where rustfmt will force types in the impl block in alphabetical order, but not for the trait block.

Could you elaborate on what you mean by this? If I recall correctly the non-default `reorder_impl_items` will both regroup const and type items within an impl, which includes sorting within those groups, but that should never be "forced".

Are you requesting a new feature that's something of an analog of `reorder_impl_items` configuration that can be used for a similar non-default sorting of types in traits?

**Comment by tisonkun** [comment]
@calebcartwright `rustfmt` will change

```rust
impl DataValue for Int64 {
    type Value = Int64Value;
    type Mutation = Int64Mutation;
}
```

to

```
impl DataValue for Int64 {
    type Mutation = Int64Mutation;
    type Value = Int64Value;
}
```

but not applied on the `trait`. I'd like to apply this rule on `trait` which means

```rust
pub trait DataValue {
    type Value;
    type Mutation;
}
```

will be formatted to

```rust
pub trait DataValue {
    type Mutation;
    type Value;
}
```

**Comment by calebcartwright** [comment]
> @calebcartwright rustfmt will change

No, rustfmt does not do this. The only way to _force_ rustfmt to make this change in impl blocks is to change the default value of the `reorder_impl_items` configuration option. You can see this in action on this [playground sample](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=b91238b43b5fbdb84a48c88dd3ef68e1).

If you are convinced that the reordering is the default behavior of rustfmt it would be great if you could provide some steps to reproduce with a minimal example, as that would be a bug.

As such, I'd echo my earlier question to you on what specifically you are asking. Are you requesting:

* that when `reorder_impl_items` is overridden to `true` that it also reorder items beyond impls/in traits?
* that the rustfmt default behavior be modified to automatically reorder items in traits? if so, are you scoping your request to types or including the reshuffling of consts, macros, etc. too?
* that a new configuration option be added, e.g. `reorder_trait_items`
* something else?

**Comment by tisonkun** [comment]
@calebcartwright you're right that `reorder_impl_items = true` is configured.

Either of the below should resolve my request:

* [x] that when `reorder_impl_items` is overridden to true that it also reorder items beyond impls/in traits?
* [x] that a new configuration option be added, e.g. `reorder_trait_items`

`reorder_trait_items` may be better.

**Comment by tisonkun** [comment]
@calebcartwright any thoughts?

**Comment by calebcartwright** [comment]
> @calebcartwright any thoughts?

I'd rather go with an explicit, new option (`reorder_trait_items`) as opposed to expanding the behavior of the existing option, especially given its name and age
